R/SOS_200-methods.R

```diff
@@ -412,159 +429,173 @@ setMethod(f = "getFeatureOfInterest", signature = signature(sos = "SOS_2.0.0", f
       warning("Got XML string, but request did not require text/xml (or subtype).")
     }
-    
+
     .parsingFunction <- sosParsers(sos)[[.parserName]]
-    
+
     if(verbose) {
       cat("[.getObservation_2.0.0] Parsing with function ")
       print(.parsingFunction)
     }
-    
+
     .obs <- .parsingFunction(obj = .response, sos = sos,
                              verbose = verbose)
-    
+
     # calculate result length vector
     if(inherits(.obs, "OmObservationCollection")) {
-      if(verbose) cat("[.getObservationById_1.0.0] Got OmObservationCollection", 
+      if(verbose) cat("[.getObservation_2.0.0] Got OmObservationCollection",
                       "... calculating length with sosResult()")
-      
+
       .result <- sosResult(.obs, bind = FALSE, coordinates = FALSE)
-      if(verbose) cat("[.getObservationById_1.0.0] result: ", toString(.result))
-      
+      if(verbose) cat("[.getObservation_2.0.0] result: ", toString(.result))
+
       .resultLength <- sapply(.result, nrow)
-      if(length(.resultLength) == 0) # nothing
-        .resultLength = 0
+      if(length(.resultLength) == 0){
+        # nothing
+        .resultLength <- 0
+      }
+    }
+    else if (is.list(.obs) && all(sapply(.obs, function(o) { class(o) == "OmOM_Observation"}))) {
+      .resultLength <- sum(sapply(.obs, function(o){ !is.null(o@result)}))
     }
     else .resultLength <- NA
-    
+
     if(verbose) {
       cat("[.getObservation_2.0.0] PARSED RESPONSE:",
           class(.obs), "\n")
       cat("[.getObservation_2.0.0] Result length(s): ",
           toString(.resultLength), "\n")
     }
-    
+
     if(is.list(.obs) && any(sapply(.obs, is.null))) {
       .countInfo <- paste("NO DATA, turn on 'verbose' for more information.")
     }
     else {
-      .countInfo <- paste(sum(.resultLength), "result values [",
+      .countInfo <- paste(sum(.resultLength), "result values")
+      if (verbose) .countInfo <- paste0(.countInfo,  " [",
                           toString(.resultLength), "].")
     }
-    
+
     .msg <- paste("[sos4R] Finished getObservation to", sos@url,
                   "\n\t--> received", length(.obs), "observation(s) having",
                   .countInfo , "\n")
     if(!is.null(.filename)) {
       .msg <- paste(.msg,
                     "[sos4R] Original document saved:", .filename, "\n")
-      
+
       .oldAttrs <- attributes(.obs)
       .newAttrs <- list(.filename)
       names(.newAttrs) <- list(sosAttributeFileName)
-      if(verbose) cat("[.getObservationById_1.0.0] Appending new attributes",
+      if(verbose) cat("[.getObservation_2.0.0] Appending new attributes",
                       toString(.newAttrs), "(names",
                       toString(names(.newAttrs)), ")\n")
-      
+
       attributes(.obs) <- c(.oldAttrs, .newAttrs)
     }
     cat(.msg)
-    
+
     # RETURN ###
     return(.obs)
   }
+  else if (.isEmptyResponse(.responseString)) {
```

Comment:
I think an empty response string is always an error, isn't it? An SOS will never send nothing. I suggest to use `stop()` here.

Reply:
The emptyResponse string represents an empty list of observations when requesting something where the SOS can not find anything, hence `stop()` would be wrong. It's a workaround because the XML processing library could not handle any XML document having only the document root element. But this is send by the SOS when the result is an empty list of observations.